tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +227,280 @@ static bool HasCanonicalDimensionNumbers(
+      size_t operand_rank = loop.index() < (max_operand_sz - 1)
+                                ? get_rank(args[loop.index()])
+                                : result_ty->getRank();
+      size_t batch_rank = operand_rank - (pre_elip + postElip);
+      // Add ("0",...,"N") where N is rank of batch into the loop
+      for (int batch_ind = 0; batch_ind < batch_rank; batch_ind++)
+        loop_dim.push_back(std::to_string(batch_ind));
+      // Add the dimension after ellipsis into the loop
+      for (int post_ch_ind = pre_elip + ellipsis.size();
+           post_ch_ind < loop.value().size(); post_ch_ind++)
+        loop_dim.push_back(std::string(1, loop.value()[post_ch_ind]));
+      batch_rank_vec.push_back(batch_rank);
+      all_loop_dim.push_back(loop_dim);
+    }
+    // Check that all ellipsis represent same rank of batch
+    if (batch_rank_vec.size() > 1) {
+      if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                      batch_rank_vec.begin())) {
+        return rewriter.notifyMatchFailure(
+            op, "Invalid Elipsis(ellipsis) within Einsum config!");
+      }
+    }
+
+    // Find all unique indices in the input and output
+    std::set<std::string> input_ind;
+    std::set<std::string> output_ind;
+    int num_of_input_operands = args.size();
+    std::vector<std::vector<std::string>> input_loops_dim_vec = {
+        all_loop_dim[0], all_loop_dim[1]};
+    for (auto input_loops_dim : input_loops_dim_vec) {
+      for (auto dim : input_loops_dim) input_ind.insert(dim);
+    }
+    for (auto dim : all_loop_dim.back()) output_ind.insert(dim);
+
+    // Check for contraction/summation indices
+    std::set<std::string> reduction_axe =
+        FindSummationAxes(input_ind, output_ind);
+
+    // Find input/output values and types.
+    auto loc = op.getLoc();
+    ValueRange inputs = args;
+
+    // Setting up Initial buffer for Output Tensor
+    Value output;
+    auto dyn_sizes = ExtractDynamicEinsumSizes(
+        rewriter, loc, args.front(), args.back(), all_loop_dim[0],
+        all_loop_dim[1], all_loop_dim[2]);
+    output = GetInitTensor(rewriter, loc, *result_ty, dyn_sizes);
+
+    // Create indexing maps.
+    // Create a 1:1 map from f:StrDimension -> affineDimension
+    int64_t nloops = input_ind.size();
+    std::unordered_map<std::string, AffineExpr> str_affine_dim_umap;
+    for (auto it : llvm::enumerate(input_ind)) {
+      str_affine_dim_umap[it.value()] = rewriter.getAffineDimExpr(it.index());
+    }
+    // For each Operands, we generate the affine map for accessing pattern.
+    // Using knowledge from einsum_config/loop of each operand in string type,
+    // we can generate the equivalent information in affine type using
+    // "getExprFromConfig". Then we can feed it into the affinemap builder
+    // where: map = mapBuilder(number_of_dims, num_constant, einsum_config,
+    // mlirCtx)
+    SmallVector<AffineMap, 4> maps;
+    for (auto it : llvm::enumerate(inputs)) {
```

Comment:
[optional] I don't know if there are conventions. I usually name it as `en` when using `llvm::enumerate`.